torch/csrc/jit/codegen/cuda/root_domain_map.h

```diff
@@ -0,0 +1,331 @@
+#include <torch/csrc/jit/codegen/cuda/iter_visitor.h>
+#include <torch/csrc/jit/codegen/cuda/utils.h>
+
+#include <torch/csrc/WindowsTorchApiMacro.h>
+
+namespace torch {
+namespace jit {
+namespace fuser {
+namespace cuda {
+
+class TORCH_CUDA_API RootDomainMap : public PolymorphicBase {
+ public:
+  //! Return a map from a producer TensorDomain to a consumer
+  //! TensorDomain
+  //!
+  //! \param producer A producer TensorDomain
+  //! \param consumer A consumer TensorDomain
+  //! \param root_dims_to_map Maps only producer root domains in this set
+  std::unordered_map<IterDomain*, IterDomain*> mapProducerToConsumer(
+      const TensorDomain* producer,
+      const TensorDomain* consumer,
+      const std::unordered_set<IterDomain*>& root_dims_to_map) const;
+
+  //! Return a map from a consumer TensorDomain to a producer
+  //! TensorDomain
+  //!
+  //! \param consumer A consumer TensorDomain
+  //! \param producer A producer TensorDomain
+  //! \param root_dims_to_map Maps only consumer root domains in this set
+  std::unordered_map<IterDomain*, IterDomain*> mapConsumerToProducer(
+      const TensorDomain* consumer,
+      const TensorDomain* producer,
+      const std::unordered_set<IterDomain*>& root_dims_to_map) const;
+
+  virtual std::ostream& print(std::ostream& os) const = 0;
+
+ protected:
+  //! Return a map between root IterDomains of a producer-consumer
+  //! pair.
+  //!
+  //! \param producer A producer TensorDomain
+  //! \param consumer A consumer TensorDomain
+  //! \param root_dims_to_map Maps only from IterDomains in this set
+  //! \param producer_to_consumer Maps from producer to consumer if true
+  virtual std::unordered_map<IterDomain*, IterDomain*> map(
+      const TensorDomain* producer,
+      const TensorDomain* consumer,
+      const std::unordered_set<IterDomain*>& root_dims_to_map,
+      bool producer_to_consumer) const = 0;
+};
+
+inline std::ostream& operator<<(std::ostream& os, const RootDomainMap& map) {
+  return map.print(os);
+}
+
+class TORCH_CUDA_API PairwiseRootDomainMap : public RootDomainMap {
+ public:
+  explicit PairwiseRootDomainMap(
+      const TensorView* producer,
+      const TensorView* consumer);
+
+  std::ostream& print(std::ostream& os) const override;
+
+ protected:
+  std::unordered_map<IterDomain*, IterDomain*> map(
+      const TensorDomain* producer,
+      const TensorDomain* consumer,
+      const std::unordered_set<IterDomain*>& root_dims_to_map,
+      bool producer_to_consumer) const override;
+
+  //! Creates mapping that does not consider the actual expression of
+  //! a producer-consumer. This is potentially unsafe.
+  PairwiseRootDomainMap() = default;
+
+ private:
+  const TensorView* producer_tv_ = nullptr;
+  const TensorView* consumer_tv_ = nullptr;
+  std::vector<bool> broadcast_flags_;
+};
+
+class TORCH_CUDA_API UnsafePairwiseRootDomainMap
+    : public PairwiseRootDomainMap {
+ public:
+  UnsafePairwiseRootDomainMap() : PairwiseRootDomainMap() {}
```

Comment:
why is this needed?

Reply:
Just for backward compatibility. For correct mapping, we need to be able to look at actual expressions used to generate TensorDomains. More specifically, `BroadcastOp::is_broadcast_dims_` needs to be consulted when the expression of a producer-consumer pair is a BroadcastOp. In the current mapping, this isn't actually done, so in some cases we generate incorrect mappings. `UnsafePairwiseRootDomainMap` doesn't check the broadcast information either, so it's potentially incorrect but should be still fine if the expression is not a BroadcastOp. So, that's why it's "unsafe".

We only have one use case of this interface, which is the RFactorReplay test. In the test, TensorDomains generated with some of the replay functions are used. The use case in the test is not for general users, but it just exercises the internal interfaces for testing. Since it is a special case, I think using the unsafe interface should be reasonable.

That said, Christian said the test could be removed as we have other RFactor tests. Then, we may not need this unsafe class at all.

I don't have any specific opinion on whether we should keep the unsafe option or not.

Reply:
Is this the one that this comment belongs to?
``` .cpp
//! Creates mapping that does not consider the actual expression of
//! a producer-consumer. This is potentially unsafe.
```
Yes, let's remove this class, and remove the replay rfactor test as it's pretty outdated at this point.

Reply:
Removed.